File: leetcode/2. problem/1277.统计全为-1-的正方形子矩阵.java
/*
 * @lc app=leetcode.cn id=1277 lang=java
 *
 * [1277] 统计全为 1 的正方形子矩阵
 */

// @lc code=start
class Solution {
    public int countSquares(int[][] matrix) {
        int n = matrix.length, m = matrix[0].length;
        int ans = 0;
        // i, j 作为正方形的右下角, dp记录边长大于1的正方形
        int[][] dp = new int[n][m];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (matrix[i][j] == 1) {
                    if (i == 0 || j == 0) {
                        dp[i][j] = 1;
                        ans++;
                        continue;
                    }
                    // 如果matrix[i][j] == 1，则此处最小有一个正方形
                    dp[i][j] = Math.min(Math.min(dp[i - 1][j], dp[i][j - 1]),
                            dp[i - 1][j - 1]) + 1;
                    ans += dp[i][j];
                }
            }
        }
        return ans;
    }
}
// @lc code=end

File: leetcode/2. problem/1277.统计全为-1-的正方形子矩阵.go
/*
 * @lc app=leetcode.cn id=1277 lang=golang
 *
 * [1277] 统计全为 1 的正方形子矩阵
 */
package main

// @lc code=start
func countSquares(matrix [][]int) int {
	var min = func(i, j int) int {
		if i > j {
			return j
		}
		return i
	}
	row, colunm := len(matrix), len(matrix[0])
	ans := 0
	dp := make([][]int, row)
	for i := 0; i < row; i++ {
		dp[i] = make([]int, colunm)
	}
	for i := 0; i < row; i++ {
		for j := 0; j < colunm; j++ {
			if matrix[i][j] == 1 {
				if i == 0 || j == 0 {
					dp[i][j] = 1
				} else {
					dp[i][j] = min(dp[i-1][j-1], min(dp[i-1][j], dp[i][j-1])) + 1
				}
				ans += dp[i][j]
			}
		}
	}

	return ans
}

// @lc code=end
